### DISCUSSION TITLE
「重学TS 2.0 」TS 面试题：TypeScript 中 any、never、unknown、null & undefined 和 void 有什么区别？

### DISCUSSION THREAD

**Discussion by semlinker** [semlinker/awesome-typescript#70] [General]
### TypeScript 中 any、never、unknown、null & undefined 和 void 有什么区别？

**Comment by wenkm** [comment]
`any` 任何类型，编译会直接跳过，写了跟js就没区别了
`never` 永不存在的类型
`unknown` 未知类型
`null` `undefined` 跟JS中一样的
`void` 没有类型，可以看做跟 `any` 相反

> 尽量不用 `any`，未知类型可以用 `unknown`，使用时用类型断言或守卫进行类型收缩

**Comment by pingan8787** [comment]
每个类型作用如下：
- `any`：表示**任意类型**，不做约束，是 TS 不写类型声明时的默认类型，编译时会直接跳过其类型检查。
```ts
let name: any = 'leo';
name = 999;
name = true;
// ...
```
详细介绍可以看文档：[any](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#any)

- `never`：表示**永不存在的类型**，TypeScript 2.0 引入，出现永不存在的返回类型的场景有”函数内执行出错，抛出异常，就不会有返回值“，比如：
```ts
// 返回never的函数必须存在无法达到的终点
function error(message: string): never {
    throw new Error(message);
}
```
`never `类型是所有类型的子类型，它可以赋值给任何类型；同时也没有类型是 `never` 的子类，除了 `never` 自身。 即使 any也不可以赋值给never。

详细介绍可以看文档：[never](https://www.typescriptlang.org/docs/handbook/2/functions.html#never)

- `unknown`：表示**未知类型**，TypeScript 3.0 引入，当不确定类型的时候可以使用，它可以被赋值给任何类型，但不能被赋值给出了 `any` 和 `unknown` 之外的类型，并且不允许执行 `unknown` 类型的变量方法（`any` 可以）。
```ts
let name: unknown = 'leo';
name = 999;
name.toString(); // Object is of type 'unknown'.(2571)
```
详细介绍可以看文档：[unknown](https://www.typescriptlang.org/docs/handbook/2/functions.html#unknown)

- `null & undefined`：这是两种 TS 的类型，默认情况下，`null` 和 `undefined` 是所有类型的子类型，可以


详细介绍可以看文档：[null and undefined](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#null-and-undefined)

- `void`：表示**无任何类型**，可以理解成和 `any` 类型相反，当函数没有任何返回值时使用，隐式返回 `undefined`，当变量声明为 `void` 类型时，该变量可以分配给 `undefined` 、`null` 或者 `void` 类型的值。
注意：如果 ts 配置文件设置了 `"strictNullChecks": false`，则分配 `null` 类型的值也报错。
详细介绍可以看文档：[void](https://www.typescriptlang.org/docs/handbook/2/functions.html#void)
